suskin/stack-template-engine
# Initial render-phase template controller

## Overview

I wanted to cut a pull request for some of the prototyping work I've been doing for template stacks, before it gets too much further and harder to follow.

This primarily prototypes the render-phase functionality for a template stack controller.

There are two phases to a template stack controller, which I am currently calling the "setup" phase and the "render" phase.

The "setup" phase occurs when the stack is first installed; this is when the stack manager or equivalent would register handlers for the different events that the template stack wants to listen for. This is not covered in this PR.

The "render" phase occurs when an event triggers one of the handlers specified by a template stack. That's what this PR shows. In the render phase, a resource engine is invoked based on the CR which triggered the render, and resources are created, updated, or deleted based on the output of the resource engine.

## Functionality

This PR shows prototypes of the following functionality:

* Starting a render phase based on a single Kind. In this case, the Kind is a `HelmChartInstall`.
* Executing a render phase using generic code which is not Kind-specific. The controller uses `unstructured.Unstructured` and hard-coded string values for the Group, Version, and Kind of the type of object being fetched.
* Executing a resource engine by starting a `Job`.
* Using helm as the resource engine.
* Passing configuration to the resource engine by pulling it off of the triggering CR. In this case, that means placing a values.yaml for the resource engine to consume.

## Tips for reading

The `render` method is the main top-level method for executing the `render` phase, so I'd recommend looking at that to get a feel for the high level flow. Also, `Reconcile` has the even more top-level flow for grabbing the claim and passing it into the render logic.

Keep in mind that this is not intended to be quite production-grade yet, though I have added many log statements to make development faster.

## Testing

I've been testing this with the following steps:

* Build the helm and kubectl helper docker images.
* Create a `StackConfiguration` by hand.
* Create a `HelmChartInstall` by hand.
* Run the controller locally by running `make run`.
* Check the resources which were created, and the Job output.
* Delete everything that was created.

## Reviewers

Since this is in my own account, I'm not seeing a way to mark people as reviewers or assignees, so I'll do it by mentioning people.

@displague @jbw976 for review
@muvaf for an fyi when we compare notes on controllers
